_posts/2020-02-26-demo-fraud-detection-2.md

```diff
@@ -0,0 +1,210 @@
+---
+layout: post
+title: "Advanced Flink Application Patterns Vol.2:
+Dynamic Updates of Application Logic"
+date: 2020-02-26T12:00:00.000Z
+authors:
+- alex:
+  name: "Alexander Fedulov"
+  twitter: "alex_fedulov"
+categories: news
+excerpt: In this series of blog posts you will learn about powerful Flink patterns for building streaming applications.
+---
+
+In the [first article](https://flink.apache.org/news/2020/01/15/demo-fraud-detection.html) of the series, we gave a high-level description of the objectives and required functionality of a Fraud Detection engine. We also described how to make data partitioning in Apache Flink customizable based on modifiable rules instead of using a hardcoded `KeysExtractor` implementation.
+
+We intentionally omitted details of how the applied rules are initialized and what possibilities exist for updating them at runtime. In this post, we will address exactly these details. You will learn how the approach to data partitioning described in [Part 1](https://flink.apache.org/news/2020/01/15/demo-fraud-detection.html) can be applied in combination with a dynamic configuration. These two patterns used together can eliminate the need to recompile the code and redeploy your Flink job for a wide range of modifications of the business logic.
+
+## Rules Broadcasting
+
+Let's first have a look at the [previously-defined](https://flink.apache.org/news/2020/01/15/demo-fraud-detection.html#dynamic-data-partitioning) data-processing pipeline:
+
+```java
+DataStream<Alert> alerts =
+    transactions
+        .process(new DynamicKeyFunction())
+        .keyBy((keyed) -> keyed.getKey());
+        .process(new DynamicAlertFunction())
+```
+
+`DynamicKeyFunction` provides dynamic data partitioning while `DynamicAlertFunction` is responsible for executing the main logic of processing transactions and sending alert messages according to defined rules.
+
+Vol.1 of this case study simplified the use case and assumed that the applied set of rules is pre-initialized and is accessible via the `List<Rules>` within `DynamicKeyFunction`.
+
+```java
+public class DynamicKeyFunction
+    extends ProcessFunction<Transaction, Keyed<Transaction, String, Integer>> {
+
+  /* Simplified */
+  List<Rule> rules = /* Rules that are initialized somehow.*/;
+  ...
+}
+```
+
+Adding rules to this list is obviously possible directly inside the code of the Flink Job at the stage of its initialization (Create a `List` object; use it's `add` method). A major drawback of doing so is that it will require recompilation of the job with each rule modification. In a real Fraud Detection system, rules are expected to change on a frequent basis, and makes this approach unacceptable from the point of view of business and operational requirements. A different approach is needed.
```

Comment:
```suggestion
Adding rules to this list is obviously possible directly inside the code of the Flink Job at the stage of its initialization (Create a `List` object; use it's `add` method). A major drawback of doing so is that it will require the recompilation of the job with each rule modification. In a real Fraud Detection system, rules are expected to change on a frequent basis, making this approach unacceptable both from a business point-of-view and its operational requirements. A different approach is thus necessary.
```

Reply:
I would like to keep "business requirements" - it is something that has specific meaning when talking about software.

Reply:
Thanks @afedulov 
I will remove the suggestion.